Gaussian Markov Random Field

nontas opened the pull request:
This PR adds `GMRFModel` and `GMRFVectorModel`. The former deals with `numpy.array` instances and the latter with Menpo's objects. This model is necessary for the implementation of Active Pictorial Structures and also provides the ability to build PCA models on graph-based sparse covariance matrices.

jabooth commented:
it would be nice here to have an additional convenience constructor for building one of our PCA Models from an existing external set of eigenvalues and eigenvectors, something along the lines of:
```py
@classmethod
PCAModel.init_from_eigenvectors(cls, evec, eval, mean)
``` 
and, for the case of instances
```py
@classmethod
PCAInstanceModel.init_from_eigenvectors(cls, evec, eval, mean, template)
```

nontas commented:
I added the
```python
PCAModel.init_from_components(cls, components, eigenvalues, mean, 
                              n_samples, centred, max_n_components=None)
```
methods in both `PCAModel` and `PCAInstanceModel`. I preferred to use the word `components` since this is what we use inside the classes.

Also, note that apart from the PCA and GMRF related changes, this PR also includes the following:
* I added some methods in `menpo/shape/graph_predefined.py` that make it easy to create some basic graphs, i.e. `empty_graph()`, `star_graph()`, `complete_graph()`, `chain_graph()` and `delaunay_graph()`. All of them have a `shape` argument that can be either a `PointCloud` or a `LandmarkGroup`, as well as a `return_pointgraph` arguments that defines whether to return a `Graph` or `PointGraph` instance.
* I added `plot_gaussian_ellipses()` in `menpo.visualize.base` which given a list of (2-dimensional) covariances and means, it plots the ellipses with an optional colourbar that ranks them with respect to their standard deviation.

patricksnape commented:
retest this please

patricksnape commented:
retest this please

nontas commented:
As soon as the tests pass, I think that this is finished. Any feedback is very welcome!

patricksnape commented on menpo/model/gmrf.py:
```diff
@@ -0,0 +1,1290 @@
+       & Pattern Recognition (CVPR), Boston, MA, USA, pp. 5435-5444, June 2015.
+    """
+    def __init__(self, samples, graph, mode='concatenation', n_components=None,
+                 single_precision=False, sparse=True, n_samples=None, bias=0,
```
The only thing we need to check is the bug you are seeing with building single_precision models!

nontas commented:
I refactored the `single_precision` parameter to `dtype`. 

I also fixed the PCA-related: In case the precision matrix (inverse covariance) `C` in `pcacov` had `np.float32` type, then the first two eigenvectors of the PCA ended up having noise. This was fixed by increasing the `eps` tolerance from ``1e-10`` to ``1e-5``.

I think this is now ready to get in.

patricksnape commented:
+1 Great work @nontas!

jabooth commented:
@nontas nice work chasing down the np.float32 bug! Just wondering, do we think that the noise thing is something that we may hit again? As in, is it proportional to the scale of the data we are taking PCA on?

nontas commented:
Maybe. However, I believe that this issue only affects the case of performing eigenanalysis on an **inverse** covariance matrix (e.g. on a GMRF's precision). Specifically, the eigenanalysis is performed as normal and then the eigenvalues are inverted and both the eigenvalues and eigenvectors are reversed, i.e.
```
eigenvalues = eigenvalues[::-1] ** -1
eigenvectors = eigenvectors[:, ::-1]
```
Thus, the small inverse eigenvalues (and eigenvectors) come to the front of the final subspace. However, if an eigenvalue was extremely small, but not small enough in order to be clipped by the `eps` value, then its inverse would be extremely high and dominate the eigenspectrum. So all we had to do is increase the `eps` threshold.

On the other hand, when performing eigenanalysis on a covariance/scatter matrix, the small eigenvalues and respective eigenvectors (which normally represent noise) will remain at the end and will probably get trimmed.

jabooth commented:
@nontas ah right I see, thanks that clears it up a lot. In that case let's just move forward with this and monitor it to see if we run into this issue again for the case of the inverse covariance matrices. +1!
